store/store.go

```diff
@@ -0,0 +1,71 @@
+package store
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	ds "github.com/ipfs/go-datastore"
+	logging "github.com/ipfs/go-log"
+	es "github.com/textileio/go-eventstore"
+	"github.com/textileio/go-eventstore/jsonpatcher"
+)
+
+const (
+	idFieldName = "ID"
+)
+
+var (
+	ErrInvalidModel = errors.New("the model is valid")
+
+	log = logging.Logger("store")
+)
+
+type Store struct {
+	datastore  ds.Datastore
+	dispatcher *es.Dispatcher
+	models     map[reflect.Type]*Model
+}
+
+func NewStore(ds ds.Datastore, dispatcher *es.Dispatcher) *Store {
+	return &Store{
+		datastore:  ds,
+		dispatcher: dispatcher,
+		models:     make(map[reflect.Type]*Model),
+	}
+}
+
+func (s *Store) RegisterJSONPatcher(name string, defaultInstance interface{}) (*Model, error) {
+	if s.alreadyRegistered(defaultInstance) {
+		return nil, fmt.Errorf("already registered model")
+	}
+
+	if !isValidModel(defaultInstance) {
+		return nil, ErrInvalidModel
+	}
+
+	eventcreator := jsonpatcher.New()
+	m := NewModel(name, defaultInstance, s.datastore, s.dispatcher, eventcreator)
+	s.models[m.valueType] = m
+	s.dispatcher.Register(m) // ToDo: find good place for reg token, with proper unregistering
```

Comment:
I'm wondering if a `Close()` method should exist in a `Store` that will unregister all the reducers from the dispatcher. Because `dispatcher.Register(..)` returns a registration token that allows the reducer to be unregistered. When this should happen? my best guess is in some `Close()` call to the `Store` that should exist?

Reply:
Agreed, de-register on Close makes sense. Do we need to de-register tokens? I can't imagine a case where they'd be used.

Reply:
Yeh, we inherited the tokens from the work I had initially done, which was copied from Flux... but we probs don't need them here, because its never going to be 'dynamic' like that.

Reply:
Not sure about it then... just kept it to not make more unilateral decisions. Maybe I'll remove the concept of the token and let the future give the hint if it is needed.